smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,232 @@
+import software.amazon.smithy.jmespath.ast.FlattenExpression
+import software.amazon.smithy.jmespath.ast.FunctionExpression
+import software.amazon.smithy.jmespath.ast.IndexExpression
+import software.amazon.smithy.jmespath.ast.LiteralExpression
+import software.amazon.smithy.jmespath.ast.MultiSelectHashExpression
+import software.amazon.smithy.jmespath.ast.MultiSelectListExpression
+import software.amazon.smithy.jmespath.ast.NotExpression
+import software.amazon.smithy.jmespath.ast.ObjectProjectionExpression
+import software.amazon.smithy.jmespath.ast.OrExpression
+import software.amazon.smithy.jmespath.ast.ProjectionExpression
+import software.amazon.smithy.jmespath.ast.SliceExpression
+import software.amazon.smithy.jmespath.ast.Subexpression
+import software.amazon.smithy.swift.codegen.SwiftWriter
+import software.amazon.smithy.swift.codegen.utils.toLowerCamelCase
+
+private val suffixSequence = sequenceOf("") + generateSequence(2) { it + 1 }.map(Int::toString) // "", "2", "3", etc.
+
+// Visits the JMESPath syntax tree, rendering the JMESPath expression into a Swift expression.
+// Smithy does not support all JMESPath expressions, and smithy-swift does not support all Smithy
+// expressions.  However, this support is sufficient to generate all waiters currently in use on AWS.
+// Use of a JMESPath feature not supported by Smithy will cause an exception at the time of code
+// generation.
+
+// Because JMESPath has no concept of non-optional, every visitor below will take an optional value
+// as input and also return an optional.
+class JMESPathVisitor(val writer: SwiftWriter) : ExpressionVisitor<String> {
+
+    // A few methods are provided here for generating unique yet still somewhat
+    // descriptive variable names when needed.
+    private val tempVars = mutableSetOf<String>()
+
+    private fun addTempVar(preferredName: String, content: String, vararg args: Any): String {
+        val name = bestTempVarName(preferredName)
+        writer.writeInline("let \$L = ", name)
+        writer.write(content, *args)
+        return name
+    }
+
+    private fun bestTempVarName(preferredName: String): String =
+        suffixSequence.map { "$preferredName$it" }.first(tempVars::add)
+
+    // Some JMESPath expressions may have their own valid JMESPath expressions
+    // within them, i.e. to map or filter.  This method is called to render
+    // those expressions.
+    private fun childBlock(forExpression: JmespathExpression): String =
+        forExpression.accept(JMESPathVisitor(writer))
+
+    // Maps the expression result in leftName into a new collection using the right expression.
+    private fun mappingBlock(right: JmespathExpression, leftName: String): String {
+        if (right is CurrentExpression) return leftName // Nothing to map
+        val outerName = bestTempVarName("projection")
+        writer.openBlock("let \$L = Optional.some((\$L ?? []).compactMap { current in", "})", outerName, leftName) {
```

Comment:
If the resulting property can be optional, then there is no need to wrap this in an Optional.

```suggestion
        writer.openBlock("let \$L = \$L?.compactMap { current in", "}", outerName, leftName) {
```